File: 백준/Gold/1753. 최단경로/최단경로.py
import heapq
import sys
from heapq import *
v,e = map(int,input().split())

k = int(input())

graph = [[] for _ in range(v+1)]


for _ in range(e):
    a,b,c=map(int, sys.stdin.readline().split())
    graph[a].append((b,c))

distance = [int(1e9)]*(v+1)

def dijkstra(start):
    q=[]
    heapq.heappush(q,(0,start))
    distance[start]=0

    while q:
        dist, now = heapq.heappop(q)

        if distance[now]<dist:
            continue

        for i in graph[now]:
            node, c = i
            cost = dist + c

            if cost<distance[node]:
                distance[node]=cost
                heapq.heappush(q,(cost, node))
dijkstra(k)

# print(distance)

for i in range(1,v+1):
    if distance[i]==int(1e9):
        print('INF')
    else:
        print(distance[i])

File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static int V,E;
	static ArrayList<ArrayList<int[]>> graph;
	static int[] distance;
	
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		V = Integer.parseInt(st.nextToken());
		E = Integer.parseInt(st.nextToken());
		int start = Integer.parseInt(br.readLine());
		distance = new int[V+1];
		
		Arrays.fill(distance, Integer.MAX_VALUE);
		
		graph = new ArrayList<>();
		for(int i=0; i<V+1; i++) {
			graph.add(new ArrayList<int[]>());
		}
		
		for(int i=0; i<E; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			int c = Integer.parseInt(st.nextToken());
			graph.get(a).add(new int[] {b,c});
		}
		
		dijstra(start);
		
		for(int i=1; i<V+1;i++) {
			if(distance[i]==Integer.MAX_VALUE)
				System.out.println("INF");
			else
				System.out.println(distance[i]);
			
		}
		
	}
	
	static void dijstra(int start) {
		
		PriorityQueue<int[]> q = new PriorityQueue<>( (o1,o2) -> o1[0]-o2[0]);
		distance[start] = 0;
		q.add(new int[] {0,start});
		
		while(!q.isEmpty()) {
			
			int[] distNode = q.poll();
			int dist = distNode[0];
			int node = distNode[1];
			
			if( dist>distance[node]) continue;
			
			
			for(int[] next : graph.get(node)) {
				int cost = dist + next[1];
				
				if(cost<distance[next[0]]) {
					distance[next[0]] = cost;
					q.add(new int[] {cost, next[0]});
				}
			}
		}
	}
	
	
	

}